Context API improvements

johnhaddon opened the pull request:
This introduces the `Context::EditableScope` class we discussed over on PR #2016. I've gone through and updated everything except GafferImage to use the new API, and based on that experience I'm pretty happy with the changes. There's less code, it's more readable, it's much harder to get things wrong, and it opens up some new opportunities for future optimisation...

andrewkaufman commented on include/GafferScene/ScenePlug.h:
```diff
@@ -146,13 +177,13 @@ class ScenePlug : public Gaffer::ValuePlug
 		/// accessing globals from within a per-location computation. It
-		/// removes unnecessary context variables which could otherwise
-		/// lead to poor cache performance.
+		/// uses GlobalScope to removes unnecessary context variables which
```
"to remove"

johnhaddon replied:
Thanks for spotting that - fixed in e937ee2.

andrewkaufman commented on src/GafferScene/ScenePlug.cpp:
```diff
@@ -271,32 +245,49 @@ const PathMatcherDataPlug *ScenePlug::setPlug() const
 	return getChild<PathMatcherDataPlug>( 7 );
 }
 
+ScenePlug::PathScope::PathScope( const Gaffer::Context *context, const ScenePath &scenePath )
```
I'm wondering if these should have convenience setters  (e.g. `PathScope::setPath` and `SetScope::setSetName`) as well as (or instead of) the constructor arg. If we've got the struct itself for convenience, might as well make it easier on the caller when they want to use it to scope a loop (as in `fullAttributes()` below).

andrewkaufman replied:
Or `update()` methods like you have on `InstanceScope`?

johnhaddon replied:
OK, sounds good - thanks for the suggestion.

johnhaddon replied:
Added in	36faaf7.

andrewkaufman commented on src/GafferScene/ScenePlug.cpp:
```diff
@@ -271,32 +245,49 @@ const PathMatcherDataPlug *ScenePlug::setPlug() const
 {
-	ContextPtr tmpContext = new Context( *Context::current(), Context::Borrowed );
-	Context::Scope scopedContext( tmpContext.get() );
+	Context::EditableScope tmpContext( Context::current() );
```
Even if we don't go with the helper setters I suggested above, maybe it'd be good to use a PathScope here anyways, to clarify the intention? It doesn't much matter for the Path case since there aren't any removes, but if someone used this as example code for a loop involving a `SetScope` they might mistakenly miss those removes.

PS. Why is this variable `tmpContext` instead of `scope`?

johnhaddon replied:
Cool, good point, I think it'll have to go hand-in-hand with the helper method and a default constructor (that doesn't set anything), because in these use cases you don't have the value to set until after you've made the scope, at least not without some contrived reordering.

andrewkaufman commented on src/GafferScene/Preview/RendererAlgo.cpp:
```diff
@@ -206,8 +204,7 @@ struct RenderSets::Updater
 	{
-		ContextPtr context = new Context( *m_context, Context::Borrowed );
-		Context::Scope scopedContext( context.get() );
+		Context::EditableScope context( m_context );
```
Same thought about using `SetScope` to clarify intentions. Actually, are we missing the removes here?

johnhaddon replied:
Cool, I'll update to use SetScope once I've added the update calls. I don't think we're actually missing the removes in this case though, because we're at the outer level of evaluation, so none of the things we'd remove should be in the context anyway.

johnhaddon replied:
Done in 4acd384.

andrewkaufman commented on src/GafferScene/SceneAlgo.cpp:
```diff
@@ -203,12 +201,11 @@ IECore::TransformPtr GafferScene::SceneAlgo::transform( const ScenePlug *scene,
 	MatrixMotionTransformPtr result = new MatrixMotionTransform();
-	ContextPtr transformContext = new Context( *Context::current(), Context::Borrowed );
-	Context::Scope scopedContext( transformContext.get() );
+	Context::EditableScope transformContext( Context::current() );
```
Should we call this variable `timeContext` since that's all we intend to vary?

johnhaddon replied:
Edited it to be `timeScope` in f41f7a9, since all in all other spots we're using `fooScope` rather than `fooContext`, and the types are `FooScope` rather than `FooContext`.

andrewkaufman commented on src/GafferScene/Preview/RendererAlgo.cpp:
```diff
@@ -157,9 +157,8 @@ class LocationTask : public tbb::task
-	GafferScene::Filter::setInputScene( c.get(), scene );
-	LocationTask<Functor> *task = new( tbb::task::allocate_root() ) LocationTask<Functor>( scene, c.get(), ScenePlug::ScenePath(), f );
+	FilterPlug::SceneScope sceneScope( Gaffer::Context::current(), scene );
+	LocationTask<Functor> *task = new( tbb::task::allocate_root() ) LocationTask<Functor>( scene, Gaffer::Context::current(), ScenePlug::ScenePath(), f );
```
Should `EditableScope` have a convenience accessor `context()` so you don't have to call `curren()` again? Not sure it matters, just a thought.

johnhaddon replied:
I did think about that, but I wasn't sure. It felt a bit weird to get a cut-down just-setting-things view onto a temporary context and then have the option of getting the whole thing back again (albeit in const form), and I wasn't sure if it would get in the way of future optimisation. Then again, the whole thing appears again from `current()` anyway, so maybe it doesn't matter?

johnhaddon replied:
I haven't done this one. One of the future optimisations I'm considering is that in some cases EditableScope might not even copy-construct the original context, but instead it would make in-place changes and then undo them at destruction. I felt like a `context()` method implies that the Scope owns a context of its own, so I'd like to give things some time to see how they pan out if it's OK...

andrewkaufman replied:
Sure

danieldresser commented:
I've been working on top of this in my image context optimization branch, and it all looks pretty good to me.

My only question is what I should call instances of EditableScope - I've left some of them named fooContext like they were before, not sure if that should be fooScope instead.  Looks like you've left some as tmpContext so far?

johnhaddon commented:
> My only question is what I should call instances of EditableScope - I've left some of them named fooContext like they were before, not sure if that should be fooScope instead. Looks like you've left some as tmpContext so far?

Yeah, I wasn't sure what to do there - I found myself wanting to call them contexts still in the case where they were just an EditableScope rather than a SetScope. I've got to do one more pass through this for Andrew today anyway, so I'll see how something more consistent feels - probably just need to detach myself from the memories of the old idiom...

johnhaddon commented:
OK, I think I've addressed all your comments @andrewkaufman, and @danieldresser I've also removed the `fooContexts` and replaced with `fooScopes`, so if you could do the same in your GafferImage branch that would be great.

andrewkaufman commented:
Looks good to me. Did you want to squash anything or shall we just wait for the tests to pass and merge as is?

johnhaddon commented:
I think I split the subsequent commits up so they'd each squash sensibly into one of the preceding commits, but if you're happy I'm happy to just merge as-is...

danieldresser commented:
Rebased as #2041 and merged there.
